crates/core/tedge_api/src/workflow/state.rs

```diff
@@ -281,6 +374,104 @@ fn json_as_string(value: &Value) -> String {
     }
 }
 
+/// A set of values to be injected/extracted into/from a [GenericCommandState]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq)]
+#[serde(try_from = "Option<Value>")]
+pub enum StateExcerpt {
+    /// A constant JSON value
+    Literal(Value),
+
+    /// A JSON path to the excerpt
+    ///
+    /// `"${.some.value.extracted.from.a.command.state}"`
+    PathExpr(String),
+
+    /// A map of named excerpts
+    ///
+    /// `{ x = "${.some.x.value}", y = "${.some.y.value}"`
+    ExcerptMap(HashMap<String, StateExcerpt>),
+
+    /// An array of excerpts
+    ///
+    /// `["${.some.x.value}", "${.some.y.value}"]`
+    ExcerptArray(Vec<StateExcerpt>),
+}
+
+impl StateExcerpt {
+    /// Extract a JSON value from the input state
+    pub fn extract_value_from(&self, input: &GenericCommandState) -> Value {
+        match self {
+            StateExcerpt::Literal(value) => value.clone(),
+            StateExcerpt::PathExpr(path) => input.extract_value(path).unwrap_or(Value::Null),
+            StateExcerpt::ExcerptMap(excerpts) => {
+                let mut values = serde_json::Map::new();
+                for (key, excerpt) in excerpts {
+                    let value = excerpt.extract_value_from(input);
+                    values.insert(key.to_string(), value);
+                }
+                Value::Object(values)
+            }
+            StateExcerpt::ExcerptArray(excerpts) => {
+                let mut values = Vec::new();
+                for excerpt in excerpts {
+                    let value = excerpt.extract_value_from(input);
+                    values.push(value);
+                }
+                Value::Array(values)
+            }
+        }
+    }
+}
+
+impl TryFrom<Option<Value>> for StateExcerpt {
+    type Error = StateExcerptError;
+
+    fn try_from(value: Option<Value>) -> Result<Self, Self::Error> {
+        match value {
+            None | Some(Value::Null) => {
+                // A mapping that change nothing
+                Ok(StateExcerpt::ExcerptMap(HashMap::new()))
+            }
+            Some(value) if value.is_object() => Ok(value.into()),
+            Some(value) => {
+                let kind = match &value {
+                    Value::Bool(_) => "bool",
+                    Value::Number(_) => "number",
+                    Value::String(_) => "string",
+                    Value::Array(_) => "array",
+                    _ => unreachable!(),
+                };
+                Err(StateExcerptError::NotAnObject {
+                    kind: kind.to_string(),
+                    value,
+                })
+            }
+        }
+    }
+}
+
+impl From<Value> for StateExcerpt {
+    fn from(value: Value) -> Self {
+        match value {
+            Value::Null => StateExcerpt::Literal(Value::Null),
+            Value::Bool(b) => StateExcerpt::Literal(Value::Bool(b)),
```

Comment:
I'm a bit confused with this inconsistency in this conversion impl compared to the impl of `From<Option<Value>>`. Literal values like boolean, number etc were error cases in that impl, but valid here? Something is not obvious.

Reply:
The difference is about the level in the tree structure of the command state payload.

- `input = true`  is rejected because the command payload as a whole must be a JSON object.
- `input.x = true` is accepted and an `x` field is added to the  command init state.